EIPS/eip-7851.md

```diff
@@ -0,0 +1,247 @@
+  - Gas is insufficient.
+  - Called via `STATICCALL` (i.e. in a read-only context).
+  - Caller is not an EOA with delegated code (prefix `0xef0100` as per [EIP-7702](./eip-7702)).
+- Updates caller's delegated code based on length:
+  - 23 bytes (`0xef0100 || address`): Appends `0x00` to deactivate private key authorization.
+  - 24 bytes (`0xef0100 || address || 0x00`): Removes last byte `0x00` to activate private key authorization.
+- Saves updated code as caller's new account code.
+
+### Transaction validation
+
+For EOAs with delegated code (begins with the prefix `0xef0100`), the following validations MUST be performed:
+
+- During the consensus rule's transaction validity check before execution: transactions signed by the private key MUST be rejected if the delegated code indicates the private key is in the deactivated state (i.e., `24` bytes long). This ensures such transactions cannot be included in blocks.
+- The transaction pool MUST implement the same validation to prevent invalid transactions from propagating across the network.
+- Any [EIP-7702](./eip-7702) authorization from an authority with a deactivated delegated code MUST be considered invalid and skipped. The gas consumption rules remain unchanged, consistent with [EIP-7702](./eip-7702).
+
+### Gas Cost
+
+The `PER_EMPTY_ACCOUNT_COST` and `PER_AUTH_BASE_COST` constants defined in [EIP-7702](./eip-7702) remain unchanged, since account code will be loaded during the authorization validation. This EIP only adds a code length check, which is a small overhead compared to existing logic.
+
+## Rationale
+
+### Using a precompiled contract
+
+Alternative methods for implementing this feature include:
+
+- Adding a new transaction type: Introducing a new transaction type could provide a mechanism to deactivate and reactivate EOA private keys. However, reactivating the private key would rely on a delegated contract as the authorizer, which complicates defining the rules for the new transaction type.
+- Deploying a regular smart contract: A regular deployed contract could track the deactivated status of each address. However, invoking this contract, which executes bytecode and accesses storage, during deactivation and reactivation, would be more expensive than using a precompiled contract. Additionally, this approach "leaks" the design of a (widely used) programming language into the Ethereum core protocol. While it poses no obvious security risks, it is not ideal from a design perspective.
+
+### In-protocol reactivation
+
+This approach ensures maximum compatibility with future migrations. EOAs can reactivate their private keys, delegate their accounts to an [EIP-7701](./eip-7701) contract, and then deactivate their private keys again. This avoids the limitations of contract upgrades. e.g. to remove legacy proxy contracts (reducing gas overhead) when upgrading to EOF contracts, one can reactivate the EOA and redelegate it to an EOF proxy contract.
```

Comment:
I see no better alternative, but this definitely sounds risky. Once a privkey has been deactivated most users would not take care to keep it secure. If it's leaked an attacker could use it after reactivation.

A powerful solution would be to enable delegation through a precompile. Then the account code could implement the required checks before "upgrading" its own code.